### PR TITLE
chore(devfile): Add first dog fooding devfile

### PR DESCRIPTION
**Pull request by sunix**
### What does this PR do?
A beginning of a dogfooding devfile to work on the devfile registry:

Provides containers and che commands to
- build in the workspace
- test the registry in a dedicated container in the workspace
- publish to surge for live update testing 

Created https://github.com/eclipse/che/issues/19305 to continue:
- investigate 'start devfile-registry' task is failing when running it from a che-theia task.
- build the offline version
- didact guide
- badge in README.md

### Screenshot/screencast of this PR

Selection_272
Selection_274

### What issues does this PR fix or reference?
N/A

### How to test this PR?
You can try the live change flow:

1. Use the factory https://workspaces.openshift.com/f?url=https://github.com/eclipse/che-devfile-registry/tree/devfile
2. make a change in `/projects/che-devfile-registry/devfiles` folder
3. run the build che-command
4. run the publish to surge.sh command
5. give a try in your own che-server changing your CheCluster CR:
    ```yaml
       devfileRegistryUrl: https://sunix-che-devfile-registry-lombok-qpd6g.surge.sh
       externalDevfileRegistry: true
    ```
7. refresh the dashboard `Get Started Page` and see your changes
5. make a change in `/projects/che-devfile-registry/devfiles` folder
6. run the build che-command
7. run the publish to surge.sh
8. refresh the dashboard `Get Started Page` and see the live changes

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: build/dev/start-devfile-registry.sh [hunk 1]
```diff
@@ -16,8 +17,10 @@ REPO_DIR=${REPO_DIR:-$DEFAULT_REPO_DIR}
...
-echo $CHE_DEVFILE_REGISTRY_URL
-/projects/che-devfile-registry/build/dockerfiles/entrypoint.sh httpd-foreground
\ No newline at end of file
+cp -rf "$BUILD_DIR"/devfiles /usr/local/apache2/htdocs/devfiles
+cp -rf "$REPO_DIR"/images/* /usr/local/apache2/htdocs/images
+CHE_DEVFILE_REGISTRY_URL=$(cat "$BUILD_DIR"/ENV_CHE_DEVFILE_REGISTRY_URL)
+export CHE_DEVFILE_REGISTRY_URL
+echo "$CHE_DEVFILE_REGISTRY_URL"
+/projects/che-devfile-registry/build/dockerfiles/entrypoint.sh echo "Starting Apache ..."
+httpd-foreground
```
Comment (ericwill) [inline build/dev/start-devfile-registry.sh:26]: New line is missing

**Review by ericwill — APPROVED** [review]
Please fix missing new-line, but other than LGTM

**Review by ibuziuk — COMMENTED** [review]
nice, is it compatible with workspaces.openshift.com? Can we have a badge in README ?

**Comment by sunix** [comment]
@ibuziuk yes this would be done in https://github.com/eclipse/che/issues/19305 to continue the work

File: build/dev/build-devfile-registry.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+rm "$BUILD_DIR" -rf;
+
+cp -rf "$REPO_DIR"/build/scripts "$BUILD_DIR"
+cp -rf "$REPO_DIR"/arbitrary-users-patch/base_images "$BUILD_DIR"
+cp -rf "$REPO_DIR"/devfiles "$BUILD_DIR"/devfiles
+cd "$BUILD_DIR"
+./check_mandatory_fields.sh devfiles
+./index.sh > devfiles/index.json
+
+DEVFILE_URL=$(sed -e 's/^"//' -e 's/"$//' <<<"$(curl "$CHE_API_INTERNAL/workspace/${CHE_WORKSPACE_ID}" -H 'Connection: keep-alive'  -H 'Accept: application/json, text/plain, */*' -H "Authorization: Bearer $CHE_MACHINE_TOKEN" | jq '.runtime.machines | .[].servers |  select(.http) | .http.url')");
```
Comment (benoitf) [inline build/dev/build-devfile-registry.sh:26]: it doesn't work with dev workspace (we don't have anymore REST API)

I am not sure why we need the URL when it's not offline registry BTW

File: build/dev/publish-devfile-registry-to-surge.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+REPO_DIR=${REPO_DIR:-$DEFAULT_REPO_DIR}
+
+rm "$SURGE_DIR" -rf;
+mkdir -p "$SURGE_DIR"/images
+cd "$SURGE_DIR"
+echo '*' > CORS
+cp -rf "$BUILD_DIR/devfiles" "$SURGE_DIR/devfiles"
+cp -rf "$REPO_DIR"/images/* "$SURGE_DIR/images"
+
+CHE_DEVFILE_REGISTRY_URL="https://$CHE_WORKSPACE_NAMESPACE-$CHE_WORKSPACE_NAME.surge.sh"
```
Comment (benoitf) [inline build/dev/publish-devfile-registry-to-surge.sh:27]: just a FYI but these env variables won't exist in dev workspace mode

File: devfile.yaml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      - sleep
+      - infinity
+    memoryLimit: 512Mi
+    type: dockerimage
+    volumes:
+      - name: images
+        containerPath: /usr/local/apache2/htdocs/images
+
+  - alias: dev
+    image: 'quay.io/eclipse/che-plugin-registry-dev:latest'
```
Comment (benoitf) [inline devfile.yaml:22]: what is the use of this image there ?
we should not use plugin-registry stuff there

File: build/dev/build-devfile-registry.sh (hunk 2 above)
Comment (sunix) [reply build/dev/build-devfile-registry.sh:26]: It is needed for the quarkus devfile which includes url to didact.md

File: devfile.yaml [hunk 5]
```diff
@@ -0,0 +1,72 @@
+apiVersion: 1.0.0
+metadata:
+  generateName: che-devfile-registry-
+
```
Comment (benoitf) [inline devfile.yaml:4]: what about enabling multi-root workspace ?

File: devfile.yaml (hunk 4 above)
Comment (sunix) [reply devfile.yaml:22]: maybe should rename this image to `che-registry-dev`: all the required deps to build devfile or plugin registry

File: devfile.yaml (hunk 5 above)
Comment (sunix) [reply devfile.yaml:4]: will give a try :) maybe add it to https://github.com/eclipse/che/issues/19305

File: build/dev/publish-devfile-registry-to-surge.sh (hunk 3 above)
Comment (sunix) [reply build/dev/publish-devfile-registry-to-surge.sh:27]: ok to rework that in https://github.com/eclipse/che/issues/19305 ?

File: devfile.yaml [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  - id: redhat/vscode-xml/latest
+    type: chePlugin
+
+  - id: redhat/vscode-yaml/latest
+    type: chePlugin
+
+  - id: eamodio/gitlens/latest
+    type: chePlugin
+    registryUrl: 'https://che-plugin-registry-main.surge.sh/v3'
```
Comment (benoitf) [inline devfile.yaml:35]: I think this nightly plug-in registry is to use that devfile in workspaces.openshift.com ?
probably there it should use the version of the plug-in embedded in the che version of CRW and not the nightly that may not work

File: build/dev/build-devfile-registry.sh (hunk 2 above)
Comment (sunix) [reply build/dev/build-devfile-registry.sh:26]: for dev workspace compatibility, ok to rework that in https://github.com/eclipse/che/issues/19305 ?

Comment (benoitf) [reply build/dev/build-devfile-registry.sh:26]: it looks like there is an oddity but yes it can be done later

File: devfile.yaml (hunk 4 above)
Comment (benoitf) [reply devfile.yaml:22]: well I think it's better to separate them
b/c we may want to publish che-plugin-registry-dev with different node version, etc.

File: devfile.yaml (hunk 6 above)
Comment (sunix) [reply devfile.yaml:35]: to me, having issues with these plugins is not a big deal. as long as it won't stop the developer to work on devfile registry.

File: devfile.yaml (hunk 4 above)
Comment (sunix) [reply devfile.yaml:22]: ok I inlude that to https://github.com/eclipse/che/issues/19305

**Comment by sunix** [comment]
@benoitf I updated https://github.com/eclipse/che/issues/19305,  let me know if it is missing something

**Review by benoitf — APPROVED** [review]

**Pull request merged**